src/bio/bio_device.c

```diff
@@ -672,3 +674,138 @@ bio_dev_list(struct bio_xs_context *xs_ctxt, d_list_t *dev_list, int *dev_cnt)
 
 	return rc;
 }
+
+int
+bio_set_led_state(struct bio_xs_context *xs_ctxt, uuid_t dev_uuid,
+		  const char *led_state, bool reset)
+{
+	struct spdk_pci_addr	pci_addr;
+	struct spdk_pci_device *pci_device;
+	struct bio_bdev	       *bio_dev;
+	struct bio_dev_info	b_info = { 0 };
+	char			addr_buf[128];
+	enum spdk_vmd_led_state current_led_state;
+	int			new_led_state;
+	int			rc = 0;
+	bool			found = false;
+
+	D_ASSERT(is_init_xstream(xs_ctxt));
+
+	bio_dev = lookup_dev_by_id(dev_uuid);
+	if (bio_dev == NULL) {
+		D_ERROR("Failed to find dev "DF_UUID"\n",
+			DP_UUID(dev_uuid));
+		return -DER_NONEXIST;
+	}
+
+	/* LED wil be reset to the orignal saved state */
+	if (reset) {
+		new_led_state = bio_dev->bb_led_state;
+		D_GOTO(skip_led_str, rc = 0);
+	}
+
+	if (led_state == NULL)
+		return -DER_INVAL;
+
+	/* Determine SPDK LED state based on led_state string */
+	if (strcasecmp(led_state, "identify") == 0)
+		new_led_state = SPDK_VMD_LED_STATE_IDENTIFY;
+	else if (strcasecmp(led_state, "on") == 0)
+		new_led_state = SPDK_VMD_LED_STATE_FAULT;
+	else if (strcasecmp(led_state, "fault") == 0)
+		new_led_state = SPDK_VMD_LED_STATE_FAULT;
+	else if (strcasecmp(led_state, "off") == 0)
+		new_led_state = SPDK_VMD_LED_STATE_OFF;
+	else {
+		D_ERROR("LED state is not valid or supported\n");
+		return -DER_NOSYS;
+	}
+
+skip_led_str:
+	rc = fill_in_traddr(&b_info, bio_dev->bb_name);
+
+	if (spdk_pci_addr_parse(&pci_addr, b_info.bdi_traddr)) {
+		D_ERROR("Unable to parse PCI address: %s\n", b_info.bdi_traddr);
+		return -DER_INVAL;
+	}
+
+	for (pci_device = spdk_pci_get_first_device(); pci_device != NULL;
+	     pci_device = spdk_pci_get_next_device(pci_device)) {
+		if (spdk_pci_addr_compare(&pci_addr, &pci_device->addr) == 0) {
+			found = true;
+			break;
+		}
+	}
+
+	if (found) {
+		if (strcmp(spdk_pci_device_get_type(pci_device), "vmd") != 0) {
+			D_ERROR("%s is not a VMD device\n", b_info.bdi_traddr);
+			return -DER_NOSYS;
+		}
+	} else {
+		D_ERROR("Unable to set led state, VMD device not found\n");
+		return -DER_INVAL;
+	}
+
+	rc = spdk_pci_addr_fmt(addr_buf, sizeof(addr_buf), &pci_device->addr);
```

Comment:
Looks this "addr_buf" is useless, it can be removed.

Reply:
yes true, it was just another check for formatting the address, but I don't end up using it. will remove